Media: Clear icon site value and setting upon deleting associated media

aduth opened the pull request:
Closes #9939

This pull request seeks to clear site icon values (both on the site object itself and the site setting) when deleting a media item which is currently assigned as the site icon. Server-side, the option is automatically cleared when the media is deleted, so we only need to recreate this behavior in the client ([related source](https://github.com/WordPress/WordPress/blob/d94530e721f7bf7ff04096a889046aaa33836f73/wp-admin/includes/class-wp-site-icon.php#L202-L216)).

__Testing Instructions:__

Testing will be made easier by use of [Redux DevTools Chrome extension](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en).

1. Navigate to [Site Settings](http://calypso.localhost:3000/settings/general)
2. Select a site
3. Assign a site icon by clicking "Change Site Icon" (select image, "Continue", crop, "Done")
  - This will set the site setting value, but does not yet change front-end icon shown
4. Click "Change Site Icon" again
5. Choose the newly created cropped image
6. Click Delete
7. In developer tools Redux tab, note that...
  - Upon the `MEDIA_DELETE` dispatched action, the diff shows that the `state.siteSettings.items[  siteId ].site_icon` was set to null
  - A `RESET_SITE_ICON` action is also dispatched, and the diff shows that the `state.sites.items[ siteId ].icon` was removed (but only if you have a site icon already assigned via non-Calypso flow)

You can also verify that deleting any other media will dispatch `MEDIA_DELETE` with no effect, and will not dispatch `RESET_SITE_ICON`.

youknowriad commented on client/post-editor/media-modal/index.jsx:
```diff
@@ -174,6 +180,11 @@ export const EditorMediaModal = React.createClass( {
+		this.props.deleteMedia( site.ID, mediaIds );
+		if ( includes( mediaIds, siteIconId ) ) {
+			this.props.resetSiteIcon( site.ID );
+		}
```
I'm a bit concerned about this code, what if we allow media deletion from elsewhere, we'll face the exact same problem. Do we have the media URLs available here, that way we could use them to compare against the siteIcon attribute in the site reducer? 

Or we could solve this differently, move this "extra" behavior to the deleteMedia action creator. You could call to requestSiteSettings action creator from there. I know it's not perfect, but maybe this could allow us to "centralize" this behavior, which seem not directly related to this component? Thoughts? I may be missing something tho :)

aduth replied:
Yeah, I had a hard time coming to a solution here. It's complicated by the fact that the data exists in many places (media objects themselves, a media ID as site setting, and a media's URL as the site icon property). But you're right that any part of the app which deletes a media would then need to be concerned with resetting the site icon, which is not great.

Similar to what I suggested at https://github.com/Automattic/wp-calypso/pull/10032#issuecomment-267093132, perhaps a middleware (the API middleware?) is the right place for these kinds of side effects. While we're not quite at the point where we want to substitute the API call for deleting media, from there we could at least check whether the media to be deleted matches the current site option.

Or, maybe since it's not related to network calls on its own, it could exist as a standalone middleware, or combined with other media middleware suggested in #9706.

cc @dmsnell in case you have thoughts on the role of middlewares / API middleware

dmsnell replied:
@aduth middleware is probably a reasonable solution for this, but it _would_ seem a bit odd with the `deleteMedia` action doing the deletion and leaving the site icon in middleware. maybe not. maybe it would actually be a quick and easy function to add to the wpcom data layer.

in essence, we should ask if deleting the media item currently set as the site icon _always_ means that the icon should be reset. if this is the case, I would expect to have some `siteIcon` reducer somewhere in the state tree listening for `mediaDelete` actions and updating in Calypso-first by resetting it. in other words, `deleteMedia` would _not_ be the genesis for resetting the site icon. this is by design in Redux. the `siteIcon` state/reducer is responsible for its own behavior and if the proposition holds, it should be listening for all relevant actions which affect it.

the middleware comes in as the _data layer_. we act as if Calypso is the raw source of truth. the site icon reducer resets its value when it see `MEDIA_DELETE` with its own media id. the middleware layer for the site icon (or site setting) endpoint also listens for this same action and will trigger an API call whenever one flies by. if the API call fails, it may very well reset the site icon to its previous id, or in this case, if the media is gone, the failure mode will do something else - set some default for example.

does this help?

youknowriad commented on client/state/media/actions.js:
```diff
@@ -0,0 +1,25 @@
+export function deleteMedia( siteId, mediaIds ) {
+	return {
+		type: MEDIA_DELETE,
+		mediaIds: castArray( mediaIds ),
```
a new Lodash tip learned for me :)

aduth commented:
A thought which might help with reconciling the icon formats is including the ID of the media item associated with the site icon on the site response, perhaps as `site.icon.media_id`. This seems like a reasonable addition anyways, as it would be useful as a pointer in case the client wants more information about the particular media item used as site icon (could include said pointer request in `site.meta.links`).

Then, as much as possible, we use the site icon ID as the source of truth, with a lookup of the associated media item if available in the current state, falling back to the site object's `icon.img` property. Ideally we would just work with the ID, rendering a `<QueryMedia id={ id } />` component to fetch the associated media in order to get its URL, but since icons are shown in the site selector, this might be unreasonably network intensive.

I'm still working on an implementation to see whether this will pan out, but figured I'd throw the idea out there.

dmsnell commented:
> but since icons are shown in the site selector, this might be unreasonably network intensive.

as an aside, this would be a good use for the API middleware. no need to refetch icons within a certain time, maybe within any given session. the middleware could what which ones have been fetched and ignore new requests for the same one.

I've been thinking about a `force` protocol overall for cases where we want to refresh for sure, which I think would be well suited for some action `meta`

aduth commented:
Ok, took me a little longer than I'd hoped to implement the approach suggested in [my previous comment](https://github.com/Automattic/wp-calypso/pull/10016#issuecomment-267364290), but is now reflected in the last handful of commits. It's a reasonable departure from the previous direction. Instead of assigning an icon URL to be tracked in site state, it's now assumed that the site items are aware of the media ID associated with the site icon by its `icon.media_id` property*. Until media is properly Reduxified, we receive the saved item into state manually after having cropped the selected image for icon. Then, when updating the site settings, the reducer for site updates its own state to use this ID. When `<SiteIcon />` renders next, it tries to look up the icon URL using the site and media state. When deleting, the sites reducer only needs to check whether the IDs of the items being deleted includes the currently assigned icon ID.

_*_ This is not yet passed by the API, and I plan to work on this next. Though for our purposes this works fairly well as-is, given fallbacks and icon updates within browser state. Only problem is if the site is received (e.g. polling) after the icon had been updated in the same session.

aduth commented:
Also, yes, the changes are quite large, and there's a decent amount of overlap with related pull requests (particularly media state). Hoping with some merging / rebasing it will come down to a more reasonable size.

aduth commented:
D2940-code

youknowriad commented on client/state/sites/reducer.js:
```diff
@@ -125,24 +127,60 @@ export function items( state = {}, action ) {
-		case SITE_ICON_SET: {
-			const { siteId, iconUrl } = action;
+		case SITE_SETTINGS_UPDATE:
+		case SITE_SETTINGS_RECEIVE: {
```
So, If I understand properly we're receiving the site Icon ID from the site settings endpoint, I'm wondering why we're storing it in this reducer, instead of relying on the siteSettings state to handle it (like other settings)?

youknowriad replied:
Nevermind, after review the API diff, I understand why you had to do that :)

aduth replied:
>So, If I understand properly we're receiving the site Icon ID from the site settings endpoint, I'm wondering why we're storing it in this reducer, instead of relying on the siteSettings state to handle it (like other settings)?
>Nevermind, after review the API diff, I understand why you had to do that :)

Yeah, with these changes, we're treating `site.icon.media_id` as the preferred source of truth for the site icon, which means we want to reconcile when incoming settings are received. Now that you mention it though, this might need to go both ways: when we receive site(s), and we know those sites in the `siteSettings` reducer, we'll probably want to update the site icon value there (though this likely would have little effect since, as I mentioned, we want to prefer using it on the site state anyways).

dmsnell replied:
> we're treating site.icon.media_id as the preferred source of truth for the site icon, which means we want to reconcile when incoming settings are received

beautiful. I love seeing these changes happen

youknowriad replied:
You mentioned in a chat that you prefer to keep the 'icon' attribute on the site endpoint to avoid extra requests. But on the same time, you're considering here `site.icon.media_id` as the preferred source of truth (which I agree with).

Maybe we should not have to care about the extra requests here (unless it's problematic, which I don't think it is), and maybe leave the DataLayer optimize this later. 

Also, if the media_id is some general information about the site, we should add it to the site endpoint directly, don't you think?

dmsnell replied:
@youknowriad there was a distinction made a long time ago between `site` and `siteSettings` on the API-side wherein `site` requests return quick data but `settings` request get data that is much slower to fetch on the server side. it's unlikely right now that this will change.

> leave the DataLayer optimize this later

I agree with this and think it's a great place to introduce the HTTP request batching, but there are steps involved there, so maybe some clever logic in the API layer can get us there in the meantime

aduth replied:
Also that `/sites/%s/settings` is read/write while `/sites/%s` is read-only.

youknowriad commented on client/components/site-icon/index.jsx:
```diff
@@ -70,6 +75,19 @@ const SiteIcon = React.createClass( {
+	// exclusively using the site ID
+	return {
+		site: getSite( state, siteId ),
+		iconUrl: getSiteIconUrl( state, siteId, imgSize )
```
This selector relies on two sub-state trees: sites and media, Shouldn't we add Queries to the component? probably a `QuerySites` since "in theory" the component could get any siteId as prop. I and guess `QueryMedia` is probably not available yet and receiveMedia is only called if we use the `SiteIconSetting` component, so I wonder how is it filled in that case?

aduth replied:
>This selector relies on two sub-state trees: sites and media, Shouldn't we add Queries to the component?

You raise a good point here. In an [earlier comment](https://github.com/Automattic/wp-calypso/pull/10016#issuecomment-267364290), I noted a worry about the number of ytihhrt from `<SiteIcon />` components since they're used for each of the user's sites in the site selector.

If we were to render a `<QuerySites />`, maybe it should be limited to the condition that the site data is not yet known (and that it's passed as a site ID, i.e. Redux)?

Regarding `<QueryMedia />`, we benefit from the fallback behavior which uses the site's `site.icon.img` property (a URL string) if the media by ID is not known. The ID only changes when updating the site icon within Calypso, at which point we've received the media explicitly (in `<SiteIconSetting />)`. I _think_ this should be enough that it would never be inaccurate, though admittedly depends on a tangled set of assumptions.

dmsnell replied:
> ytihhrt

??

aduth replied:
>>ytihhrt
>
>??

😆 

Corrected to "requests".

dmsnell replied:
> If we were to render a <QuerySites />, maybe it should be limited to the condition that the site data is not yet known (and that it's passed as a site ID, i.e. Redux)?

Why is this? Is it for data usage? If so, maybe we should be able to do a new fetch just for the site info data. This would be trivial in the new API data layer since we could trap some action such as `SITE_ICON_REQUEST` and make an API call presumably to the same site settings endpoint from where it comes but then limiting the response fields to only the site icon data.

aduth replied:
>Why is this?

Mainly because otherwise every time I open my site selector, it would issue about 30 requests for each site shown, even if we already have data for those sites from `/me/sites`. We have this implied assumption in Calypso that sites data is expected to be universally requested via `/me/sites`, not specifically through any one query component. Not that this is very good, mind you, but the current situation nonetheless.

aduth commented on client/components/site-icon/index.jsx:
```diff
@@ -70,6 +75,19 @@ const SiteIcon = React.createClass( {
+export default connect( ( state, { site, siteId, imgSize } ) => {
+	// Until sites state is completely within Redux, we provide compatibility
+	// in cases where site object is passed to use the icon.img property as URL
+	if ( site ) {
```
>Just asking, one could refactor the parent component and provide a "redux" site instead no?

In some cases, yes. But only for sites we know to be in state, which at the moment only includes the currently selected site. Since site icons are shown in the site selector, we can't yet use Redux state there.

youknowriad commented:
While testing the API side, I noticed that for WPCOM Sites using Site Icon (and not blavatar) don't have the `icon` property on the `/sites/%s` endpoint. Should wa add a `<QuerySiteSettings>` to make sure this usecase is covered by this PR?

aduth commented:
Ah, sorry I was a bit late in pushing my rebase @youknowriad . Not too much has changed at a high level, but specifically:

- Reorganized, squashed, and shuffled changes around in commit history
- Opted to remove `keyedReducer` for now since I do not want to persist media (yet) and this is not currently possible with the utility. Instead reimplemented reducers manually keying by `siteId`.
- Moved selectors to global selectors directory
- Split off `getSiteIconId` into distinct selector. This is a tad granular for my liking, but I think is important to ensure we have consistent retrieval of site ID.

To expand on the last point, I had mentioned that we may need to add handlers to site settings to update the `site_icon` properties for sites it tracks, but in retrospect, if we are to treat `sites` as the preferred source for icon ID, this is unnecessary. Using a selector should discourage developers from trying to access the site setting's `site_icon` directly.

aduth commented on client/state/site-settings/reducer.js:
```diff
@@ -64,7 +66,21 @@ export const items = createReducer( {}, {
 		}
-	} )
+	} ),
+	[ MEDIA_DELETE ]: ( state, { siteId, mediaIds } ) => {
```
Hmm, given my note at https://github.com/Automattic/wp-calypso/pull/10016#issuecomment-269004720, maybe we don't even need to handle this.

youknowriad replied:
If I understand correctly, you're saying that we always rely on the `site` object to retrieve the siteIcon. Thus, we should not worry about resetting it on the siteSettings? Is this assumption correct? Maybe we should consider dropping the selector fallback then, since the `site-settings` value could be incorrect (even if in the current workflow, it can never happen since the site object will be available in this case) ?

aduth replied:
Hmm, you raise a good point. The fallback exists in case we ever receive site settings before the site itself. But if that were ever to occur, we would still want to monitor for media deletions even if we only knew the site settings. But we still do not need to account for `SITE_RECEIVE` or `SITES_RECEIVE` in the site settings reducer because at that point we can assume that the fallback will no longer be used.

Because of this, I'll plan to drop 15cb3f5 in a rebase.

youknowriad commented on client/state/selectors/get-media-item.js:
```diff
@@ -0,0 +1,16 @@
+/**
```
Great to see selectors making their way to the global directory

youknowriad commented:
I tested after the rebase and it's still working properly
